Source/Parallelization/WarpXRegrid.cpp

```diff
@@ -228,10 +248,112 @@ WarpX::RemakeLevel (int lev, Real time, const BoxArray& ba, const DistributionMa
             costs[lev]->setVal(0.0);
         }
 
+    if (costs_heuristic[lev] != nullptr) {
+            costs_heuristic[lev].reset(new MultiFab(costs_heuristic[lev]->boxArray(), dm, 1, 0));
+            costs_heuristic[lev]->setVal(0.0);
+        }
+
         SetDistributionMap(lev, dm);
     }
     else
     {
         amrex::Abort("RemakeLevel: to be implemented");
     }
 }
+
+void
+WarpX::ComputeCostsHeuristic (int lev)
+{
+    MultiFab* cost = WarpX::getCostsHeuristic(lev);
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+    auto nSpecies = mypc.nSpecies();
+    const IntVect& ng = current_fp[lev][0]->nGrowVect();
+
+    // Species loop
+    for (int i_s = 0; i_s < nSpecies; ++i_s)
+    {
+        auto & myspc = mypc.GetParticleContainer(i_s);
+
+        // Particle loop
+        for (WarpXParIter pti(myspc, lev); pti.isValid(); ++pti)
+        {
+            if (cost)
+            {
+                const Box& tbx = pti.tilebox();
+                const long np = pti.numParticles();
+                Real wt = (costs_heuristic_prtl_wt*(1.0*np)/tbx.d_numPts()); // prtl wt per (work) cell
+                Array4<Real> const& costarr = cost->array(pti);
+                amrex::ParallelFor(
+                                   tbx,
+                                   [=] AMREX_GPU_DEVICE (int i, int j, int k)
+                                   {
+                                       costarr(i,j,k) += wt;
+                                   }
+                                  );
+            }
+        }
+    }
+
+    // Cell loop
+    for (MFIter mfi(*cost); mfi.isValid(); ++mfi)
+    {
+        if (cost) {
+        const Box& tbx = mfi.tilebox();
+    const Box& grown_tbx = mfi.growntilebox(ng);
+    int tbx_size = 1;
+    int grown_tbx_size = 1;
+
+    // Compute n_cells for each box
+    for (auto el : tbx.size())
+    {
+        tbx_size *= el;
+    }
+
+    for (auto el : grown_tbx.size())
+    {
+        grown_tbx_size *= el;
+    }
+
+        Real wt = (costs_heuristic_cell_wt*1.0*grown_tbx_size/tbx_size); // cell wt per (work) cell
+        Array4<Real> const& costarr = cost->array(mfi);
+        amrex::ParallelFor(
+                           tbx,
+                           [=] AMREX_GPU_DEVICE (int i, int j, int k)
+                           {
+                               costarr(i,j,k) += wt;
+                           }
+                          );
+        }
+    }
```

Comment:
I think this could be replaced by
```c++
    int n_cells = 0;
    Multifab* Ex = Efield_fp[lev][0].get();
    for (MFIter mfi(*Ex, false); mfi.isValid(); ++mfi){
        const Box& gbx = mfi.growntilebox();
        n_cells += gbx.numPts();
    }
```
Then, we could do
```c++
    cost[lev] = costs_heuristic_particles_wt * n_particles +
                costs_heuristic_cells_wt * n_cells;
```

Reply:
Replaced.  As a double check, I confirm that I compute identical distribution mappings whether using (now) vector of reals, or the (old) calculation with MultiFabs.